fedora_messaging/tests/unit/test_cli.py

```diff
@@ -76,17 +77,219 @@ def test_bad_conf(self, mock_consume):
+        result = self.runner.invoke(cli.cli, ["consume"])
+        mock_importlib.import_module.not_called()
+        mock_consume.assert_not_called()
+        self.assertIn(
+            "No bindings are defined in the configuration file"
+            " and none were provided as arguments!",
+            result.output,
+        )
+        self.assertEqual(1, result.exit_code)
+
+    @mock.patch.dict("fedora_messaging.config.conf", {"amqp_url": "a", "bindings": "b"})
+    @mock.patch("fedora_messaging.cli.importlib")
+    @mock.patch("fedora_messaging.cli.api.consume")
+    def test_good_cli_callable(self, mock_consume, mock_importlib):
+        """Assert providing a callable via the CLI works."""
+        cli_options = {"callback": "mod:callable"}
+        mock_mod_with_callable = mock.Mock(spec=["callable"])
+        mock_importlib.import_module.return_value = mock_mod_with_callable
+        result = self.runner.invoke(
+            cli.cli, ["consume", "--callback=" + cli_options["callback"]]
+        )
+        mock_importlib.import_module.called_once_with("mod")
+        mock_consume.assert_called_once_with(mock_mod_with_callable.callable, "b")
+        self.assertEqual(0, result.exit_code)
+
+    @mock.patch.dict(
+        "fedora_messaging.config.conf",
+        {"amqp_url": "a", "bindings": "b", "callback": None},
+    )
+    @mock.patch("fedora_messaging.cli.importlib")
+    @mock.patch("fedora_messaging.cli.api.consume")
+    def test_missing_cli_and_conf_callable(self, mock_consume, mock_importlib):
+        """Assert missing callable via cli and in conf is reported."""
+        mock_mod_with_callable = mock.Mock(spec=["callable"])
+        mock_importlib.import_module.return_value = mock_mod_with_callable
+        result = self.runner.invoke(cli.cli, ["consume"])
+        mock_importlib.import_module.not_called()
+        mock_consume.assert_not_called()
+        self.assertIn(
+            '"callback" must be the Python path to a' " callable to consume",
+            result.output,
+        )
+        self.assertEqual(1, result.exit_code)
+
+    @mock.patch.dict("fedora_messaging.config.conf", {"amqp_url": "a", "bindings": "b"})
+    @mock.patch("fedora_messaging.cli.importlib")
+    @mock.patch("fedora_messaging.cli.api.consume")
+    def test_cli_callable_wrong_format(self, mock_consume, mock_importlib):
+        """Assert a wrong callable format is reported."""
+        cli_options = {"callback": "modcallable"}
+        error_message = "not enough values to unpack (expected 2, got 1)"
+        mock_mod_with_callable = mock.Mock(spec=["callable"])
+        mock_importlib.import_module.return_value = mock_mod_with_callable
+        result = self.runner.invoke(
+            cli.cli, ["consume", "--callback=" + cli_options["callback"]]
+        )
+        mock_importlib.import_module.not_called()
+        mock_consume.assert_not_called()
+        self.assertIn(
+            "Unable to parse the callback path ({}); the "
```

Comment:
Unfortunately it looks like the unpacking error message changes from 3.4 to 3.5. One way to fix this would be to change the assertions in some way to deal with it, but maybe displaying the nitty-gritty error to the user isn't useful.

Perhaps it'd be better to change this error message to just say ``'Unable to parse the callback path "{provided_callback_path}"; the expected format is "my_package.module:callable_object"' ``. Having the provided path and the expected path format right next to each other should be enough for someone to figure out what's wrong.

Reply:
I notice this after "push", in Travis output. I forgot to test in pyrhon 2.7.
There is a second way to test it(for the future testing and for education purposes):
line -> callback_path = callback or config.conf["callback"]
can return Mock, which can in line
module, cls = callback_path.strip().split(":")
throw ValueError with custom err message.
But I think that your solution will improve portability.
What do you think?

Reply:
Let's change the error message :+1: